{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Downloading the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example usage:\n",
    "accession_numbers = [\n",
    "    \"SRR23891364\", \"SRR23891365\", \"SRR23891368\",\n",
    "    \"SRR23891369\", \"SRR23891371\", \"SRR23891373\",\n",
    "    \"SRR23891374\", \"SRR23891375\", \"SRR23891376\",\n",
    "    \"SRR23891377\", \"SRR23891378\", \"SRR23891363\",\n",
    "    \"SRR23891366\", \"SRR23891367\", \"SRR23891370\",\n",
    "    \"SRR23891372\"\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def download_data(accession_numbers, output_directory):\n",
    "    for accession_number in accession_numbers:\n",
    "        command = f\"fastq-dump -O {output_directory} {accession_number}\"\n",
    "        os.system(command)\n",
    "\n",
    "output_directory = \"SRA_data/\"\n",
    "download_data(accession_numbers, output_directory)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for accession_number in accession_numbers:\n",
    "    !gzip SRA_data/{accession_number}.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Trimming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for accession_number in accession_numbers:\n",
    "    # Create the output directory if it doesn't exist\n",
    "    trimmed_output_dir='no_length/processed/trimmed/'\n",
    "    os.makedirs(trimmed_output_dir, exist_ok=True)\n",
    "    input_path = os.path.join(\"SRA_data\", f\"{accession_number}.fastq.gz\")\n",
    "    output_path = os.path.join(trimmed_output_dir, accession_number)\n",
    "    !trim_galore --quality 5 --output_dir $output_path $input_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating Index for non-mrna "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runMode genomeGenerate --genomeDir no_length/non-mrna/non-mRNA_ref/ --genomeFastaFiles no_length/non-mrna/non-mRNAfasta/C_albicans_SC5314_A22_current_other_features_no_introns.fasta --runThreadN 16\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 12:08:01 ..... started STAR run\n",
      "Dec 06 12:08:01 ... starting to generate Genome files\n",
      "!!!!! WARNING: --genomeSAindexNbases 14 is too large for the genome size=806909, which may cause seg-fault at the mapping step. Re-run genome generation with recommended --genomeSAindexNbases 8\n",
      "Dec 06 12:08:01 ... starting to sort Suffix Array. This may take a long time...\n",
      "Dec 06 12:08:01 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Dec 06 12:08:04 ... loading chunks from disk, packing SA...\n",
      "Dec 06 12:08:04 ... finished generating suffix array\n",
      "Dec 06 12:08:04 ... generating Suffix Array index\n",
      "Dec 06 12:08:06 ... completed Suffix Array index\n",
      "Dec 06 12:08:06 ... writing Genome to disk ...\n",
      "Dec 06 12:08:06 ... writing Suffix Array to disk ...\n",
      "Dec 06 12:08:06 ... writing SAindex to disk\n",
      "Dec 06 12:08:08 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "os.makedirs(\"no_length/non-mrna/non-mRNA_ref/\", exist_ok=True)\n",
    "!STAR --runMode genomeGenerate \\\n",
    "--genomeDir no_length/non-mrna/non-mRNA_ref/ \\\n",
    "--genomeFastaFiles no_length/non-mrna/non-mRNAfasta/C_albicans_SC5314_A22_current_other_features_no_introns.fasta \\\n",
    "--runThreadN 16"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are some annotation problems in the current GTF file, hence it is always recommended to check whether your gtf file is properly formatted or not. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed 's/\"\"/\"/g' Gtf/DatasetS3.gtf > Gtf/DatasetS3_1.gtf\n",
    "!sed 's/\"transcript_id/transcript_id/; s/;\"/;/' Gtf/DatasetS3_1.gtf> Gtf/DatasetS3_final.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating index for mrna "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --runMode genomeGenerate --genomeDir no_length/mrna/ref/ --genomeSAindexNbases 11 --sjdbGTFfeatureExon exon --genomeFastaFiles no_length/Assembly_22_fasta/C_albicans_SC5314_A22_current_chromosomes.fasta --sjdbGTFfile no_length/Gtf/DatasetS3_final.gtf\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 12:12:05 ..... started STAR run\n",
      "Dec 06 12:12:05 ... starting to generate Genome files\n",
      "Dec 06 12:12:05 ..... processing annotations GTF\n",
      "Dec 06 12:12:05 ... starting to sort Suffix Array. This may take a long time...\n",
      "Dec 06 12:12:06 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Dec 06 12:12:50 ... loading chunks from disk, packing SA...\n",
      "Dec 06 12:12:50 ... finished generating suffix array\n",
      "Dec 06 12:12:50 ... generating Suffix Array index\n",
      "Dec 06 12:12:52 ... completed Suffix Array index\n",
      "Dec 06 12:12:52 ..... inserting junctions into the genome indices\n",
      "Dec 06 12:12:52 ... writing Genome to disk ...\n",
      "Dec 06 12:12:52 ... writing Suffix Array to disk ...\n",
      "Dec 06 12:12:52 ... writing SAindex to disk\n",
      "Dec 06 12:12:52 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "os.makedirs('no_length/mrna/ref', exist_ok=True)\n",
    "# build genome index using GTF file\n",
    "!STAR --runThreadN 16 \\\n",
    "     --runMode genomeGenerate \\\n",
    "     --genomeDir no_length/mrna/ref/ \\\n",
    "     --genomeSAindexNbases 11 \\\n",
    "     --sjdbGTFfeatureExon exon\\\n",
    "     --genomeFastaFiles no_length/Assembly_22_fasta/C_albicans_SC5314_A22_current_chromosomes.fasta \\\n",
    "     --sjdbGTFfile no_length/Gtf/DatasetS3_final.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "non-mRNA mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['no_length/processed/trimmed/SRR23891364/SRR23891364_trimmed.fq.gz', 'no_length/processed/trimmed/SRR23891365/SRR23891365_trimmed.fq.gz', 'no_length/processed/trimmed/SRR23891368/SRR23891368_trimmed.fq.gz', 'no_length/processed/trimmed/SRR23891369/SRR23891369_trimmed.fq.gz', 'no_length/processed/trimmed/SRR23891373/SRR23891373_trimmed.fq.gz', 'no_length/processed/trimmed/SRR23891375/SRR23891375_trimmed.fq.gz', 'no_length/processed/trimmed/SRR23891376/SRR23891376_trimmed.fq.gz', 'no_length/processed/trimmed/SRR23891377/SRR23891377_trimmed.fq.gz', 'no_length/processed/trimmed/SRR23891378/SRR23891378_trimmed.fq.gz']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "trimmed_fastq_directory=f'no_length/processed/trimmed/'\n",
    "# Define the paths to your trimmed FASTQ files\n",
    "suffix= '_trimmed.fq.gz'\n",
    "trimmed_fastq_files = [os.path.join(trimmed_fastq_directory,str(accession_number), str(accession_number)+suffix) for accession_number in accession_numbers]\n",
    "print(trimmed_fastq_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/non-mrna/non-mRNA_ref --outFilterMismatchNmax 2 --outFileNamePrefix no_length/non-mrna/mapped_reads/bams/SRR23891375.map_rRNA_ --readFilesIn no_length/processed/trimmed/SRR23891375/SRR23891375_trimmed.fq.gz --outReadsUnmapped Fastx --readFilesCommand zcat --limitBAMsortRAM 9659450608--outSAMtype BAM SortedByCoordinate\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 15:58:11 ..... started STAR run\n",
      "Dec 06 15:58:11 ..... loading genome\n",
      "Dec 06 15:58:13 ..... started mapping\n",
      "Dec 06 16:17:21 ..... finished mapping\n",
      "Dec 06 16:17:21 ..... finished successfully\n",
      "Mapping completed for SRR23891375. Output BAM: no_length/non-mrna/mapped_reads/bams/SRR23891375.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/non-mrna/non-mRNA_ref --outFilterMismatchNmax 2 --outFileNamePrefix no_length/non-mrna/mapped_reads/bams/SRR23891376.map_rRNA_ --readFilesIn no_length/processed/trimmed/SRR23891376/SRR23891376_trimmed.fq.gz --outReadsUnmapped Fastx --readFilesCommand zcat --limitBAMsortRAM 9659450608--outSAMtype BAM SortedByCoordinate\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 16:17:22 ..... started STAR run\n",
      "Dec 06 16:17:22 ..... loading genome\n",
      "Dec 06 16:17:23 ..... started mapping\n",
      "Dec 06 16:41:33 ..... finished mapping\n",
      "Dec 06 16:41:33 ..... finished successfully\n",
      "Mapping completed for SRR23891376. Output BAM: no_length/non-mrna/mapped_reads/bams/SRR23891376.bam\n",
      "Read mapping process completed.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "accession_numbers = [\n",
    "    \"SRR23891367\", \"SRR23891370\", \"SRR23891372\" \n",
    "] \n",
    "\n",
    "trimmed_fastq_directory=f'no_length/processed/trimmed/'\n",
    "\n",
    "# Define the paths to your trimmed FASTQ files\n",
    "suffix= '_trimmed.fq.gz'\n",
    "trimmed_fastq_files = [os.path.join(trimmed_fastq_directory,str(accession_number), str(accession_number)+suffix) for accession_number in accession_numbers]\n",
    "\n",
    "\n",
    "# Define the output directories\n",
    "output_bam_dir = 'no_length/non-mrna/mapped_reads/bams/'\n",
    "output_unmapped_fastq_dir = 'no_length/non-mrna/mapped_reads/unmapped_fastq/'\n",
    "output_starlogs_dir = 'no_length/non-mrna/mapped_reads/starlogs/'\n",
    "# temp_dir = '/tmp/mergedrRNA/'\n",
    "\n",
    "# Create the temporary directory\n",
    "# os.makedirs(temp_dir, exist_ok=True)\n",
    "# Create the output directories if they don't exist\n",
    "os.makedirs(output_bam_dir, exist_ok=True)\n",
    "os.makedirs(output_unmapped_fastq_dir, exist_ok=True)\n",
    "os.makedirs(output_starlogs_dir, exist_ok=True)\n",
    "\n",
    "# Rule: map_star_rRNA\n",
    "for trimmed_fastq_file in trimmed_fastq_files:\n",
    "    sample_name = os.path.basename(trimmed_fastq_file).split('_trimmed.fq.gz')[0]\n",
    "    output_bam = os.path.join(output_bam_dir, f'{sample_name}.bam')\n",
    "    output_unmapped_fastq = os.path.join(output_unmapped_fastq_dir, f'{sample_name}.unmapped_rRNA.fastq.gz')\n",
    "    output_starlogs = os.path.join(output_starlogs_dir, f'{sample_name}_starlogs')\n",
    "\n",
    "    # STAR command for read mapping\n",
    "    star_cmd = (\n",
    "        f'STAR --runThreadN 16 '\n",
    "        f'--genomeDir no_length/non-mrna/non-mRNA_ref '\n",
    "        f'--outFilterMismatchNmax 2 '\n",
    "        f'--outFileNamePrefix {output_bam.replace(\".bam\", \".map_rRNA_\")} '\n",
    "        f'--readFilesIn {trimmed_fastq_file} '\n",
    "        f'--outReadsUnmapped Fastx '\n",
    "        f'--readFilesCommand zcat ' \n",
    "        f'--outSAMtype BAM SortedByCoordinate '\n",
    "        f'--limitBAMsortRAM 9659450608'\n",
    "    )\n",
    "\n",
    "    # Run the STAR command\n",
    "    os.system(star_cmd)\n",
    "\n",
    "    print(f\"Mapping completed for {sample_name}. Output BAM: {output_bam}\")\n",
    "\n",
    "# Print a message indicating the completion\n",
    "print(\"Read mapping process completed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 46 files and 1 in-memory blocks...\n",
      "[E::aux_parse] Incomplete aux field\n",
      "[W::sam_read1_sam] Parse error at line 32205770\n",
      "samtools sort: truncated file. Aborting\n"
     ]
    }
   ],
   "source": [
    "# accession_numbers = [\n",
    "#     \"SRR23891375\", \"SRR23891376\" \n",
    "# ] \n",
    "\n",
    "# for accession in accession_numbers:\n",
    "#     !samtools sort no_length/non-mrna/mapped_reads/bams/{accession}.map_rRNA_Aligned.out.sam -o no_length/non-mrna/mapped_reads/bams/{accession}.map_rRNA_Aligned.sortedByCoord.out.bam\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mv: cannot stat 'no_length/non-mrna/mapped_reads/bams/SRR23891364.map_rRNA_Unmapped.out.mate1': No such file or directory\n",
      "mv: cannot stat 'no_length/non-mrna/mapped_reads/bams/SRR23891365.map_rRNA_Unmapped.out.mate1': No such file or directory\n",
      "mv: cannot stat 'no_length/non-mrna/mapped_reads/bams/SRR23891368.map_rRNA_Unmapped.out.mate1': No such file or directory\n",
      "mv: cannot stat 'no_length/non-mrna/mapped_reads/bams/SRR23891369.map_rRNA_Unmapped.out.mate1': No such file or directory\n",
      "mv: cannot stat 'no_length/non-mrna/mapped_reads/bams/SRR23891373.map_rRNA_Unmapped.out.mate1': No such file or directory\n",
      "mv: cannot stat 'no_length/non-mrna/mapped_reads/bams/SRR23891375.map_rRNA_Unmapped.out.mate1': No such file or directory\n",
      "mv: cannot stat 'no_length/non-mrna/mapped_reads/bams/SRR23891377.map_rRNA_Unmapped.out.mate1': No such file or directory\n",
      "mv: cannot stat 'no_length/non-mrna/mapped_reads/bams/SRR23891378.map_rRNA_Unmapped.out.mate1': No such file or directory\n",
      "mv: cannot stat 'no_length/non-mrna/mapped_reads/bams/SRR23891366.map_rRNA_Unmapped.out.mate1': No such file or directory\n"
     ]
    }
   ],
   "source": [
    "for accession_number in accession_numbers:\n",
    "    !mv 'no_length/non-mrna/mapped_reads/bams/{accession_number}.map_rRNA_Unmapped.out.mate1' \"no_length/non-mrna/mapped_reads/unmapped_fastq/{accession_number}.unmapped_rRNA.fastq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\n"
     ]
    }
   ],
   "source": [
    "for accession_number in accession_numbers:\n",
    "    !gzip \"no_length/non-mrna/mapped_reads/unmapped_fastq/{accession_number}.unmapped_rRNA.fastq\"\n",
    "    !gzip \"no_length/non-mrna/mapped_reads/bams/{accession_number}.map_rRNA_Aligned.sortedByCoord.out.bam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891364.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891365.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891368.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891369.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891373.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891375.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891377.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891378.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891366.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891370.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891372.unmapped_rRNA.fastq.gz', 'no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891367.unmapped_rRNA.fastq.gz']\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "# Define the paths to your trimmed FASTQ files\n",
    "unmapped_fastq_directory = 'no_length/non-mrna/mapped_reads/unmapped_fastq/'\n",
    "suffix= '.unmapped_rRNA.fastq.gz'\n",
    "# Get a list of all unmapped FastQ.gz files in the directory\n",
    "fastq_files = [os.path.join(unmapped_fastq_directory,str(accession_number)+suffix) for accession_number in accession_numbers]\n",
    "print(fastq_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891367_unmapped_rRNA.fastq.gz\n",
    "no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891367.unmapped_rRNA.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "mRNA mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891364.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891364.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 19:50:21 ..... started STAR run\n",
      "Dec 06 19:50:21 ..... loading genome\n",
      "Dec 06 19:50:21 ..... started mapping\n",
      "Dec 06 19:52:40 ..... finished mapping\n",
      "Dec 06 19:52:40 ..... started sorting BAM\n",
      "Dec 06 19:53:53 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891364.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891364.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891364.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891365.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891365.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 19:53:53 ..... started STAR run\n",
      "Dec 06 19:53:53 ..... loading genome\n",
      "Dec 06 19:53:54 ..... started mapping\n",
      "Dec 06 19:56:10 ..... finished mapping\n",
      "Dec 06 19:56:10 ..... started sorting BAM\n",
      "Dec 06 19:56:54 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891365.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891365.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891365.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891368.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891368.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 19:56:55 ..... started STAR run\n",
      "Dec 06 19:56:55 ..... loading genome\n",
      "Dec 06 19:56:55 ..... started mapping\n",
      "Dec 06 21:02:36 ..... finished mapping\n",
      "Dec 06 21:02:37 ..... started sorting BAM\n",
      "Dec 06 21:03:00 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891368.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891368.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891368.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891369.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891369.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 21:03:01 ..... started STAR run\n",
      "Dec 06 21:03:01 ..... loading genome\n",
      "Dec 06 21:03:01 ..... started mapping\n",
      "Dec 06 21:35:01 ..... finished mapping\n",
      "Dec 06 21:35:01 ..... started sorting BAM\n",
      "Dec 06 21:35:18 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891369.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891369.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891369.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891373.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891373.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 21:35:18 ..... started STAR run\n",
      "Dec 06 21:35:18 ..... loading genome\n",
      "Dec 06 21:35:18 ..... started mapping\n",
      "Dec 06 21:37:21 ..... finished mapping\n",
      "Dec 06 21:37:21 ..... started sorting BAM\n",
      "Dec 06 21:37:44 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891373.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891373.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891373.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891375.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891375.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 21:37:44 ..... started STAR run\n",
      "Dec 06 21:37:44 ..... loading genome\n",
      "Dec 06 21:37:45 ..... started mapping\n",
      "Dec 06 22:00:46 ..... finished mapping\n",
      "Dec 06 22:00:46 ..... started sorting BAM\n",
      "Dec 06 22:01:12 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891375.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891375.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891375.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891377.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891377.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 22:01:13 ..... started STAR run\n",
      "Dec 06 22:01:13 ..... loading genome\n",
      "Dec 06 22:01:13 ..... started mapping\n",
      "Dec 06 22:24:48 ..... finished mapping\n",
      "Dec 06 22:24:48 ..... started sorting BAM\n",
      "Dec 06 22:25:05 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891377.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891377.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891377.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891378.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891378.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 22:25:05 ..... started STAR run\n",
      "Dec 06 22:25:05 ..... loading genome\n",
      "Dec 06 22:25:05 ..... started mapping\n",
      "Dec 06 22:29:56 ..... finished mapping\n",
      "Dec 06 22:29:56 ..... started sorting BAM\n",
      "Dec 06 22:30:07 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891378.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891378.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891378.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891366.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891366.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 22:30:08 ..... started STAR run\n",
      "Dec 06 22:30:08 ..... loading genome\n",
      "Dec 06 22:30:08 ..... started mapping\n",
      "Dec 06 22:31:49 ..... finished mapping\n",
      "Dec 06 22:31:49 ..... started sorting BAM\n",
      "Dec 06 22:32:13 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891366.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891366.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891366.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891370.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891370.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 22:32:13 ..... started STAR run\n",
      "Dec 06 22:32:13 ..... loading genome\n",
      "Dec 06 22:32:13 ..... started mapping\n",
      "Dec 06 22:37:39 ..... finished mapping\n",
      "Dec 06 22:37:39 ..... started sorting BAM\n",
      "Dec 06 22:37:46 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891370.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891370.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891370.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891372.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891372.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 22:37:46 ..... started STAR run\n",
      "Dec 06 22:37:46 ..... loading genome\n",
      "Dec 06 22:37:47 ..... started mapping\n",
      "Dec 06 22:39:57 ..... finished mapping\n",
      "Dec 06 22:39:57 ..... started sorting BAM\n",
      "Dec 06 22:40:23 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891372.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891372.bam\": No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891372.bam\n",
      "\t/usr/lib/rna-star/bin/STAR-avx2 --runThreadN 16 --genomeDir no_length/mrna/ref --outFilterMismatchNmax 2 --readFilesCommand zcat --outFileNamePrefix no_length/mapped_mRNA/bams_star/SRR23891367.map_mRNA --readFilesIn no_length/non-mrna/mapped_reads/unmapped_fastq/SRR23891367.unmapped_rRNA.fastq.gz --alignEndsType EndToEnd --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 18659450608\n",
      "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
      "Dec 06 22:40:23 ..... started STAR run\n",
      "Dec 06 22:40:23 ..... loading genome\n",
      "Dec 06 22:40:23 ..... started mapping\n",
      "Dec 06 23:03:29 ..... finished mapping\n",
      "Dec 06 23:03:29 ..... started sorting BAM\n",
      "Dec 06 23:03:42 ..... finished successfully\n",
      "Mapped BAM file: no_length/mapped_mRNA/bams_star/SRR23891367.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"no_length/mapped_mRNA/bams_star/SRR23891367.bam\" : No such file or directory\n",
      "samtools index: failed to open \"no_length/mapped_mRNA/bams_star/SRR23891367.bam\": No such file or directory\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "# Define the paths to your input FASTQ files\n",
    "unmapped_fastq_directory = 'no_length/non-mrna/mapped_reads/unmapped_fastq/'\n",
    "suffix= '.unmapped_rRNA.fastq.gz'\n",
    "# Get a list of all unmapped FastQ.gz files in the directory\n",
    "fastq_files = [os.path.join(unmapped_fastq_directory,str(accession_number)+suffix) for accession_number in accession_numbers]\n",
    "\n",
    "\n",
    "# Define the STAR index path\n",
    "star_index = 'no_length/mrna/ref'\n",
    "\n",
    "# Define the output directories\n",
    "output_dir_mapped = 'no_length/mapped_mRNA/bams_star/'\n",
    "# output_dir_unmapped = 'no_length/unmapped_mRNA/fastq/'\n",
    "output_dir_logs = 'no_length/mapped_mRNA/starlogs/'\n",
    "\n",
    "# Create the output directories if they don't exist\n",
    "os.makedirs(output_dir_mapped, exist_ok=True)\n",
    "# os.makedirs(output_dir_unmapped, exist_ok=True)\n",
    "os.makedirs(output_dir_logs, exist_ok=True)\n",
    "\n",
    "# Rule: map_star_mRNA\n",
    "# Map the unmapped RNA reads to the mRNA index using STAR\n",
    "for fastq_file in fastq_files:\n",
    "    sample_name = os.path.basename(fastq_file).split('.')[0]\n",
    "    output_bam = os.path.join(output_dir_mapped, f'{sample_name}.bam')\n",
    "    # output_fastq = os.path.join(output_dir_unmapped, f'{sample_name}.unmapped_mRNA.fastq.gz')\n",
    "    output_logs = os.path.join(output_dir_logs, f'{sample_name}.map_mRNALog')\n",
    "\n",
    "    star_cmd = f'''\n",
    "        STAR --runThreadN 16 \\\n",
    "        --genomeDir {star_index} \\\n",
    "        --outFilterMismatchNmax 2 \\\n",
    "        --readFilesCommand zcat \\\n",
    "        --outFileNamePrefix {output_bam.replace(\".bam\", \".map_mRNA\")} --readFilesIn {fastq_file} \\\n",
    "        --alignEndsType EndToEnd \\\n",
    "        --outSAMtype BAM SortedByCoordinate \\\n",
    "        --limitBAMsortRAM 18659450608 \\\n",
    "    '''\n",
    "    \n",
    "    os.system(star_cmd)\n",
    "\n",
    "#     # Print the paths of the output files\n",
    "#     print(\"Mapped BAM file:\", output_bam)\n",
    "#     # print(\"Unmapped FASTQ file:\", output_fastq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for accession_number in accession_numbers:\n",
    "#     !mv 'no_length/mapped_mRNA/bams_star/{accession_number}_unmapped_rRNA.map_mRNAUnmapped.out.mate1' \"no_length/unmapped_mRNA/fastq/{accession_number}_unmapped_rRNA.fastq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for accession_number in accession_numbers:\n",
    "    # !gzip \"no_length/unmapped_mRNA/fastq/{accession_number}_unmapped_rRNA.fastq\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter Unique Mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "\n",
    "# Set the paths\n",
    "input_dir = 'no_length/mapped_mRNA/bams_star/'\n",
    "output_dir = 'no_length/mapped_mRNA/bams_NH1/'\n",
    "tmp_dir = '/tmp'\n",
    "\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "os.makedirs(tmp_dir, exist_ok=True)\n",
    "\n",
    "\n",
    "# Create a sample_list          \n",
    "sample_list = [f'{number}.map_mRNAAligned.sortedByCoord.out' for number in accession_numbers]\n",
    "\n",
    "# Iterate over samples\n",
    "for sample in sample_list:\n",
    "    input_bam = os.path.join(input_dir, f'{sample}.bam')\n",
    "    output_bam = os.path.join(output_dir, f'{sample}.NH1.bam')\n",
    "    output_unsorted_bam = f'{output_bam}.unsorted'\n",
    "\n",
    "    # Rule: filter_uniq_mapping\n",
    "    cmd = f\"\"\"\n",
    "        bamtools filter -tag NH:1 -mapQuality \">=255\" -mapQuality \"<256\" -in {input_bam} -out {output_bam}.unsorted &&\n",
    "        samtools sort -T {tmp_dir}/{sample}_uniq_mapping -o {output_bam} {output_bam}.unsorted &&\n",
    "        samtools index {output_bam} &&\n",
    "        rm {output_bam}.unsorted\n",
    "    \"\"\"\n",
    "\n",
    "    try:\n",
    "        # Run the command\n",
    "        subprocess.run(cmd, shell=True, check=True)\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error occurred while processing {sample}: {e}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter Twice Mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 8 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 7 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 6 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 5 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 7 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 7 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 5 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 8 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 7 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "\n",
    "# Set the paths\n",
    "input_dir = 'no_length/mapped_mRNA/bams_star/'\n",
    "output_dir = 'no_length/mapped_mRNA/bams_NH2/'\n",
    "tmp_dir = '/tmp'\n",
    "\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "os.makedirs(tmp_dir, exist_ok=True)\n",
    "\n",
    "\n",
    "# List of samples\n",
    "sample_list = [f'{number}.map_mRNAAligned.sortedByCoord.out' for number in accession_numbers] \n",
    "\n",
    "# Iterate over samples\n",
    "for sample in sample_list:\n",
    "    input_bam = os.path.join(input_dir, f'{sample}.bam')\n",
    "    output_bam = os.path.join(output_dir, f'{sample}.NH2.bam')\n",
    "    output_unsorted_bam = f'{output_bam}.unsorted'\n",
    "\n",
    "    # Rule: filter_twice_mapping\n",
    "    # Based on https://groups.google.com/d/msg/rna-star/FhVk4f61Vcg/osx6DLTpAgAJ\n",
    "    cmd = f\"\"\"\n",
    "        bamtools filter -tag NH:2 -mapQuality \"<255\" -in {input_bam} -out {output_unsorted_bam} &&\n",
    "        samtools sort -T {tmp_dir}/{sample}_twice_mapping -o {output_bam} {output_unsorted_bam} &&\n",
    "        samtools index {output_bam} &&\n",
    "        rm {output_unsorted_bam}\n",
    "    \"\"\"\n",
    "\n",
    "    try:\n",
    "        # Run the command\n",
    "        subprocess.run(cmd, shell=True, check=True)\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error occurred while processing {sample}: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collapsing NH2 to NH1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pip install pysam\n",
    "import pysam  \n",
    "import sys\n",
    "\n",
    "input_dir = 'no_length/mapped_mRNA/bams_NH2/'\n",
    "output_dir = 'no_length/mapped_mRNA/bams_NH2ToNH1/'\n",
    "threads = 16  # Replace with the actual number of threads\n",
    "tmp_dir = '/tmp'\n",
    "\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "os.makedirs(tmp_dir, exist_ok=True)\n",
    "\n",
    "def collapse_nh2_to_nh1(input_bam, output_bam):\n",
    "    # Open the input BAM file\n",
    "    with pysam.AlignmentFile(input_bam, \"rb\") as in_bam:\n",
    "        # Open the output BAM file\n",
    "        with pysam.AlignmentFile(output_bam, \"wb\", header=in_bam.header) as out_bam:\n",
    "            # Keep track of read names to only output one alignment per read\n",
    "            seen_reads = set()\n",
    "\n",
    "            # Iterate through each read in the input BAM file\n",
    "            for read in in_bam:\n",
    "                # Check if the read has the NH tag equal to 2\n",
    "                if read.get_tag(\"NH\") == 2:\n",
    "                    read_name = read.query_name\n",
    "\n",
    "                    # If this read has not been seen before, output it\n",
    "                    if read_name not in seen_reads:\n",
    "                        # Change the NH tag to 1\n",
    "                        read.set_tag(\"NH\", 1)\n",
    "                        out_bam.write(read)\n",
    "                        seen_reads.add(read_name)\n",
    "    return\n",
    "\n",
    "\n",
    "# List of samples\n",
    "sample_list = [f'{number}.map_mRNAAligned.sortedByCoord.out' for number in accession_numbers] # Replace with your actual sample names\n",
    "\n",
    "for sample in sample_list:\n",
    "\n",
    "    input_bam = os.path.join(input_dir, f'{sample}.NH2.bam')\n",
    "    output_bam = os.path.join(output_dir, f'{sample}.NH2ToNH1.bam.unsorted')\n",
    "    collapse_nh2_to_nh1(input_bam, output_bam)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sorting NH2ToNH1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "\n",
    "# Set the paths\n",
    "input_dir = 'no_length/mapped_mRNA/bams_NH2/'\n",
    "output_dir = 'no_length/mapped_mRNA/bams_NH2ToNH1/'\n",
    "threads = 16  \n",
    "tmp_dir = '/tmp'\n",
    "\n",
    "os.makedirs(output_dir, exist_ok=True)                 \n",
    "os.makedirs(tmp_dir, exist_ok=True)\n",
    "\n",
    "# List all BAM files in the input directory\n",
    "sample_list = [f'{number}.map_mRNAAligned.sortedByCoord.out' for number in accession_numbers] # Replace with your actual sample names\n",
    "\n",
    "\n",
    "for sample in sample_list:\n",
    "\n",
    "    input_bam = os.path.join(input_dir, f'{sample}.NH2.bam')\n",
    "    output_bam = os.path.join(output_dir, f'{sample}.NH2ToNH1.bam')\n",
    "    output_unsorted_bam = f'{output_bam}.unsorted'\n",
    "\n",
    "    # Rule: filter_twice_mapping\n",
    "    # Based on https://groups.google.com/d/msg/rna-star/FhVk4f61Vcg/osx6DLTpAgAJ\n",
    "    cmd = f\"\"\"\n",
    "        samtools index {input_bam} &&\n",
    "        samtools sort -T {tmp_dir}/{sample}_collapse -o {output_bam} {output_bam}.unsorted &&\n",
    "        samtools index {output_bam} &&\n",
    "        rm {output_bam}.unsorted\n",
    "    \"\"\"\n",
    "\n",
    "    try:\n",
    "        # Run the command\n",
    "        subprocess.run(cmd, shell=True, check=True)\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error occurred while processing {sample_name}: {e}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merging NH2ToNH1 with NH1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "\n",
    "# Set the paths\n",
    "input_dir_nh2nh1 = 'no_length/mapped_mRNA/bams_NH2ToNH1/'\n",
    "input_dir_nh1 = 'no_length/mapped_mRNA/bams_NH1/'\n",
    "output_dir_collapsed = 'no_length/mapped_mRNA/bams_collapsed/'\n",
    "threads = 16  # Replace with the actual number of threads\n",
    "tmp_dir = '/tmp'\n",
    "\n",
    "os.makedirs(output_dir_collapsed, exist_ok=True)\n",
    "os.makedirs(tmp_dir, exist_ok=True)                                        \n",
    "\n",
    "# List all NH2ToNH1 BAM files in the input directory\n",
    "sample_list = [f'{number}.map_mRNAAligned.sortedByCoord.out' for number in accession_numbers] # Replace with your actual sample names\n",
    "\n",
    "# Iterate through each NH2ToNH1 BAM file\n",
    "for sample in sample_list:\n",
    "    \n",
    "    # Construct the full paths for NH2ToNH1 and NH1 BAM files\n",
    "    input_bam_nh2nh1 = os.path.join(input_dir_nh2nh1, f'{sample}.NH2ToNH1.bam')\n",
    "    input_bam_nh1 = os.path.join(input_dir_nh1, f'{sample}.NH1.bam')\n",
    "    output_bam_collapsed = os.path.join(output_dir_collapsed, f'{sample}.bam')\n",
    "   \n",
    "\n",
    "    # Rule: merge_NH2NH1_NH1\n",
    "    cmd = f\"\"\"\n",
    "        bamtools merge -in {input_bam_nh2nh1} -in {input_bam_nh1} -out {output_bam_collapsed}.unsorted &&\n",
    "        samtools sort -T {tmp_dir}/{sample}_collapsed -o {output_bam_collapsed} {output_bam_collapsed}.unsorted &&\n",
    "        samtools index {output_bam_collapsed} &&\n",
    "        rm {output_bam_collapsed}.unsorted\n",
    "    \"\"\"\n",
    "\n",
    "    try:\n",
    "        # Run the command\n",
    "        subprocess.run(cmd, shell=True, check=True)\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error occurred while processing {sample_name}: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FeatureCount\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Separating GTF into hapA and hapB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "!grep 'gene_id \".*_A\"' no_length/Gtf/DatasetS3_final.gtf > no_length/Gtf/GTF_hapA.gtf\n",
    "!grep 'gene_id \".*_B\"' no_length/Gtf/DatasetS3_final.gtf > no_length/Gtf/GTF_hapB.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we also have gene-ids which do not have hapA and hapB! Hence, now I'll be removing those. \n",
    "Also, the gene-id \"MSTRG\" is not needed for further processes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "!grep -v 'gene_id \"MSTRG' no_length/Gtf/DatasetS3_final.gtf > no_length/Gtf/GTF_noM.gtf\n",
    "!grep -v 'gene_id \".*_[AB]\"' no_length/Gtf/GTF_noM.gtf > no_length/Gtf/GTF_noAB.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using featureCounts with hapA gtf :-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891364.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891364.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891364.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891364.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m21163869\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m16534912 (78.1%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.16 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891364.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891365.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891365.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891365.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891365.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m18957876\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m14739276 (77.7%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.14 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891365.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891368.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891368.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891368.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891368.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m15770101\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m12441457 (78.9%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.10 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891368.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891369.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891369.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891369.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891369.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m14150472\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m11148341 (78.8%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.09 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891369.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891371.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891371.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891371.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891371.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m15383353\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m12062933 (78.4%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.10 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891371.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891373.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891373.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891373.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891373.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m20022878\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m15515696 (77.5%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.15 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891373.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891374.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891374.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891374.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891374.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m17049864\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m12948706 (75.9%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.11 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891374.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891375.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891375.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891375.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891375.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m19877103\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m13150627 (66.2%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.12 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891375.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891376.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891376.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891376.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891376.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m21396867\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m16236053 (75.9%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.15 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891376.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891377.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891377.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891377.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891377.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m14470176\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m11610059 (80.2%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.09 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891377.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891378.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891378.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891378.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891378.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m9091982\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m6752930 (74.3%)\u001b[0m \u001b[0m                     ||\n",
      "||    Running time : \u001b[36m0.08 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891378.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891363.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891363.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891363.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891363.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m15736139\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m12100939 (76.9%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.11 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891363.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891366.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891366.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891366.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891366.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m23406520\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m18280294 (78.1%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.15 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891366.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891367.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891367.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891367.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891367.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m8840144\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m5360599 (60.6%)\u001b[0m \u001b[0m                     ||\n",
      "||    Running time : \u001b[36m0.06 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891367.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891370.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891370.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891370.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891370.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m5862900\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m4778639 (81.5%)\u001b[0m \u001b[0m                     ||\n",
      "||    Running time : \u001b[36m0.04 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891370.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891372.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapA_SRR23891372.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapA_SRR23891372.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapA.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapA\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapA.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6185\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6185\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891372.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m20585447\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m15961842 (77.5%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.17 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pA/counts_hapA_SRR23891372.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# GTF file\n",
    "gtf_file = \"no_length/Gtf/GTF_hapA.gtf\"\n",
    "\n",
    "# Output directory for featureCounts results                      \n",
    "output_counts_dir = \"no_length/reads_gene/hapA\"\n",
    "\n",
    "# Create the output directory if it doesn't exist\n",
    "!mkdir -p $output_counts_dir\n",
    "\n",
    "# Iterate through accession numbers\n",
    "for accession_number in accession_numbers:                        \n",
    "    # Input BAM file\n",
    "    input_bam = f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\"\n",
    "    \n",
    "    # Output counts file\n",
    "    output_counts = f\"{output_counts_dir}/counts_hapA_{accession_number}.txt\"\n",
    "    # Run featureCounts command\n",
    "    !featureCounts -g gene_id -a $gtf_file -o $output_counts $input_bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now working with the hapB gtf file:-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891364.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891364.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891364.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891364.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m21163869\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m847451 (4.0%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.15 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891364.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891365.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891365.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891365.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891365.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m18957876\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m844639 (4.5%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.14 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891365.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891368.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891368.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891368.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891368.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m15770101\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m533105 (3.4%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.12 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891368.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891369.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891369.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891369.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891369.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m14150472\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m490206 (3.5%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.09 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891369.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891371.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891371.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891371.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891371.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m15383353\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m605375 (3.9%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.09 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891371.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891373.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891373.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891373.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891373.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m20022878\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m838136 (4.2%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.15 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891373.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891374.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891374.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891374.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891374.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m17049864\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m672061 (3.9%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.11 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891374.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891375.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891375.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891375.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891375.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m19877103\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m657867 (3.3%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.12 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891375.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891376.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891376.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891376.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891376.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m21396867\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m691590 (3.2%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.13 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891376.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891377.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891377.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891377.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891377.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m14470176\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m504944 (3.5%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.08 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891377.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891378.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891378.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891378.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891378.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m9091982\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m308813 (3.4%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.06 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891378.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891363.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891363.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891363.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891363.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m15736139\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m556429 (3.5%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.10 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891363.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891366.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891366.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891366.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891366.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m23406520\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m802233 (3.4%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.13 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891366.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891367.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891367.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891367.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891367.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m8840144\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m261576 (3.0%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.05 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891367.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891370.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891370.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891370.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891370.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m5862900\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m215554 (3.7%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.04 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891370.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891372.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_hapB_SRR23891372.txt\u001b[0m \u001b[0m                     ||\n",
      "||                 Summary : \u001b[36mcounts_hapB_SRR23891372.txt.summary\u001b[0m \u001b[0m             ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_hapB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/hapB\u001b[0m \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_hapB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m6184\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6184\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m8\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891372.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m20585447\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m803514 (3.9%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.13 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/ha \u001b[0m ||\n",
      "|| pB/counts_hapB_SRR23891372.txt.summary\" \u001b[0m                                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# GTF file\n",
    "gtf_file = \"no_length/Gtf/GTF_hapB.gtf\"\n",
    "\n",
    "# Output directory for featureCounts results                      \n",
    "output_counts_dir = \"no_length/reads_gene/hapB\"\n",
    "\n",
    "# Create the output directory if it doesn't exist\n",
    "!mkdir -p $output_counts_dir\n",
    "\n",
    "# Iterate through accession numbers\n",
    "for accession_number in accession_numbers:                        \n",
    "    # Input BAM file\n",
    "    input_bam = f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\"\n",
    "    \n",
    "    # Output counts file\n",
    "    output_counts = f\"{output_counts_dir}/counts_hapB_{accession_number}.txt\"\n",
    "    # Run featureCounts command\n",
    "    !featureCounts -g gene_id -a $gtf_file -o $output_counts $input_bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now working with the no_hapAB gtf file:-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891364.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891364.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891364.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891364.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m21163869\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m454998 (2.1%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.13 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891364.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891365.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891365.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891365.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891365.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m18957876\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m432156 (2.3%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.12 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891365.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891368.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891368.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891368.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891368.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m15770101\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m42075 (0.3%)\u001b[0m \u001b[0m                        ||\n",
      "||    Running time : \u001b[36m0.10 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891368.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891369.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891369.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891369.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891369.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m14150472\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m132507 (0.9%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.08 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891369.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891371.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891371.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891371.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891371.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m15383353\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m564919 (3.7%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.09 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891371.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891373.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891373.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891373.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891373.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m20022878\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m742290 (3.7%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.12 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891373.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891374.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891374.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891374.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891374.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m17049864\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m713361 (4.2%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.11 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891374.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891375.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891375.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891375.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891375.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m19877103\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m425009 (2.1%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.11 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891375.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891376.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891376.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891376.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891376.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m21396867\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m343954 (1.6%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.13 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891376.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891377.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891377.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891377.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891377.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m14470176\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m115554 (0.8%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.08 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891377.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891378.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891378.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891378.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891378.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m9091982\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m294652 (3.2%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.06 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891378.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891363.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891363.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891363.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891363.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m15736139\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m386175 (2.5%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.10 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891363.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891366.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891366.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891366.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891366.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m23406520\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m576362 (2.5%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.13 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891366.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891367.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891367.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891367.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891367.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m8840144\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m64024 (0.7%)\u001b[0m \u001b[0m                        ||\n",
      "||    Running time : \u001b[36m0.05 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891367.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891370.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891370.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891370.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891370.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m5862900\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m31029 (0.5%)\u001b[0m \u001b[0m                        ||\n",
      "||    Running time : \u001b[36m0.04 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891370.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.3\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                           \u001b[36mSRR23891372.map_mRNAAligned.sortedByCoord.ou\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcounts_nohapAB_SRR23891372.txt\u001b[0m \u001b[0m                  ||\n",
      "||                 Summary : \u001b[36mcounts_nohapAB_SRR23891372.txt.summary\u001b[0m \u001b[0m          ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||              Annotation : \u001b[36mGTF_noAB.gtf (GTF)\u001b[0m \u001b[0m                              ||\n",
      "||      Dir for temp files : \u001b[36mno_length/reads_gene/nohapAB\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file GTF_noAB.gtf \u001b[0m... \u001b[0m                                     ||\n",
      "||    Features : \u001b[36m20\u001b[0m \u001b[0m                                                          ||\n",
      "||    Meta-features : \u001b[36m20\u001b[0m \u001b[0m                                                     ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file SRR23891372.map_mRNAAligned.sortedByCoord.out.bam... \u001b[0m     ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m20585447\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m740522 (3.6%)\u001b[0m \u001b[0m                       ||\n",
      "||    Running time : \u001b[36m0.13 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Write the final count table. \u001b[0m                                              ||\n",
      "|| Write the read assignment summary. \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"no_length/reads_gene/no \u001b[0m ||\n",
      "|| hapAB/counts_nohapAB_SRR23891372.txt.summary\" \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# GTF file\n",
    "gtf_file = \"no_length/Gtf/GTF_noAB.gtf\"\n",
    "\n",
    "# Output directory for featureCounts results                      \n",
    "output_counts_dir = \"no_length/reads_gene/nohapAB\"\n",
    "\n",
    "# Create the output directory if it doesn't exist\n",
    "!mkdir -p $output_counts_dir\n",
    "\n",
    "# Iterate through accession numbers\n",
    "for accession_number in accession_numbers:                        \n",
    "    # Input BAM file\n",
    "    input_bam = f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\"\n",
    "    \n",
    "    # Output counts file\n",
    "    output_counts = f\"{output_counts_dir}/counts_nohapAB_{accession_number}.txt\"\n",
    "    # Run featureCounts command\n",
    "    !featureCounts -g gene_id -a $gtf_file -o $output_counts $input_bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Converting txt to csv files for ease of processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "for accession_number in accession_numbers:\n",
    "\n",
    "    out_dir1= \"no_length/reads_gene/hapA/csv\"\n",
    "    os.makedirs(out_dir1, exist_ok=True)\n",
    "    out_dir2= \"no_length/reads_gene/hapB/csv\"\n",
    "    os.makedirs(out_dir2, exist_ok=True)\n",
    "    out_dir3= \"no_length/reads_gene/nohapAB/csv\"\n",
    "    os.makedirs(out_dir3, exist_ok=True)\n",
    "\n",
    "    df = pd.read_csv(f\"no_length/reads_gene/hapA/counts_hapA_{accession_number}.txt\", skiprows=1)\n",
    "    df.to_csv(f'{out_dir1}/counts_hapA_{accession_number}.csv', index=False)\n",
    "\n",
    "    df1 = pd.read_csv(f\"no_length/reads_gene/hapB/counts_hapB_{accession_number}.txt\", skiprows=1)\n",
    "    df1.to_csv(f'{out_dir2}/counts_hapB_{accession_number}.csv', index=False)\n",
    "\n",
    "    df2 = pd.read_csv(f'no_length/reads_gene/nohapAB/counts_nohapAB_{accession_number}.txt', skiprows=1)\n",
    "    df2.to_csv(f'{out_dir3}/counts_nohapAB_{accession_number}.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we want to add the counts for hapA and hapB gene-ids and then concatinate the resulting counts with the nohapAB counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "dfs=[]\n",
    "\n",
    "\n",
    "for accession_number in accession_numbers:      \n",
    "    # Read the files for hapA and hapB \n",
    "    df1 = pd.read_csv(f\"no_length/reads_gene/hapA/csv/counts_hapA_{accession_number}.csv\", sep=\"\\t\") \n",
    "    df2 = pd.read_csv(f\"no_length/reads_gene/hapB/csv/counts_hapB_{accession_number}.csv\", sep=\"\\t\")\n",
    "\n",
    "\n",
    "    # Remove the last two characters from the Geneid column              \n",
    "\n",
    "    df1[\"Geneid\"] = df1[\"Geneid\"].str.slice(0, -2)\n",
    "    df2[\"Geneid\"] = df2[\"Geneid\"].str.slice(0, -2)\n",
    "        \n",
    "    # df_len= df1[[\"Geneid\",\"Length\"]]  ## exracting length for further DeSEQ and TE analysis\n",
    "\n",
    "    # Extract the columns Geneid and reads\n",
    "    df1 = df1[[\"Geneid\", f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\"]]\n",
    "    df2 = df2[[\"Geneid\", f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\"]]\n",
    "\n",
    "  \n",
    "\n",
    "    # Rename the reads column\n",
    "    df1 = df1.rename(columns={f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\": f\"reads_{accession_number}_x\"})\n",
    "    df2 = df2.rename(columns={f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\": f\"reads_{accession_number}_y\"})\n",
    "\n",
    "    # Merge the two dataframes on Geneid\n",
    "    df = pd.merge(df1, df2, on=\"Geneid\", how=\"inner\")\n",
    "\n",
    "    # Convert the \"reads\" columns to numeric\n",
    "    df[f\"reads_{accession_number}_x\"] = pd.to_numeric(df[f\"reads_{accession_number}_x\"], errors='coerce')\n",
    "    df[f\"reads_{accession_number}_y\"] = pd.to_numeric(df[f\"reads_{accession_number}_y\"], errors='coerce')\n",
    "\n",
    "    # Sum the reads columns\n",
    "    df[f\"reads_{accession_number}\"] = df[f\"reads_{accession_number}_x\"] + df[f\"reads_{accession_number}_y\"]\n",
    "\n",
    "    # Drop the redundant reads columns\n",
    "    df = df.drop(columns=[f\"reads_{accession_number}_x\", f\"reads_{accession_number}_y\"])\n",
    "\n",
    "    # Append the dataframe to the list\n",
    "    dfs.append(df)\n",
    "\n",
    "df_final = pd.concat(dfs, axis=1)\n",
    "\n",
    "# Drop the duplicate Geneid columns\n",
    "df_final = df_final.loc[:, ~df_final.columns.duplicated()]\n",
    "# df_len.to_csv(\"no_length/reads_gene/length_AB.csv\", sep=\"\\t\", index=False)\n",
    "# Write the dataframe to a new file\n",
    "df_final.to_csv(\"no_length/reads_gene/counts_sum.csv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we want to concatinate the above csv file with the nohapAB csv files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfss=[]\n",
    "\n",
    "for accession_number in accession_numbers:\n",
    "    df= pd.read_csv(f\"no_length/reads_gene/nohapAB/csv/counts_nohapAB_{accession_number}.csv\", sep=\"\\t\")\n",
    "\n",
    "    # df_len=df[[\"Geneid\", \"Length\"]]\n",
    "\n",
    "    df = df[[\"Geneid\", f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\"]]\n",
    "    \n",
    "\n",
    "    # Rename the reads column\n",
    "    df = df.rename(columns={f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\": f'reads_{accession_number}'})\n",
    "\n",
    "    dfss.append(df)\n",
    "\n",
    "df_final = pd.concat(dfss, axis=1)\n",
    "df_final = df_final.loc[:, ~df_final.columns.duplicated()]\n",
    "\n",
    "# Write the dataframe to a new file\n",
    "df_final.to_csv(\"no_length/reads_gene/counts_sum_nohapAB.csv\", sep=\"\\t\", index=False)\n",
    "# df_len.to_csv(\"no_length/reads_gene/length_noAB.csv\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         Geneid  reads_SRR23891364  reads_SRR23891365  reads_SRR23891368  \\\n",
      "0     C1_00010W                  0                  0                 81   \n",
      "1     C1_00020C                  0                  0                 12   \n",
      "2     C1_00030C                  0                  0                  0   \n",
      "3     C1_00040W                 67                113                 40   \n",
      "4     C1_00050C                  0                  0                  5   \n",
      "...         ...                ...                ...                ...   \n",
      "6194  CM_00340W              11561              13861                222   \n",
      "6195  CM_00370W               5110               4366                494   \n",
      "6196  CM_00380W              36826              35523               2923   \n",
      "6197  CM_00440W              20734              22667               2157   \n",
      "6198  CM_00520W                  8                 33                  1   \n",
      "\n",
      "      reads_SRR23891369  reads_SRR23891371  reads_SRR23891373  \\\n",
      "0                     2                 46                  0   \n",
      "1                     0                  0                  1   \n",
      "2                     0                  0                  0   \n",
      "3                    41                 67                 75   \n",
      "4                     0                  1                  3   \n",
      "...                 ...                ...                ...   \n",
      "6194               2699               8315              23273   \n",
      "6195                952               4629               8091   \n",
      "6196              15153              44839              61888   \n",
      "6197               8269              29915              47928   \n",
      "6198                  0                 10                  1   \n",
      "\n",
      "      reads_SRR23891374  reads_SRR23891375  reads_SRR23891376  \\\n",
      "0                    14                 48                 42   \n",
      "1                    22                  0                  1   \n",
      "2                     0                  0                  1   \n",
      "3                    26                  5                  5   \n",
      "4                     0                  4                  3   \n",
      "...                 ...                ...                ...   \n",
      "6194              20232               5877               2115   \n",
      "6195               8864               3837               1898   \n",
      "6196              65196              31342              28022   \n",
      "6197              40862              20887              18542   \n",
      "6198                 20                  1                  2   \n",
      "\n",
      "      reads_SRR23891377  reads_SRR23891378  reads_SRR23891363  \\\n",
      "0                     7                  8                  8   \n",
      "1                     2                  1                  0   \n",
      "2                     3                  1                  0   \n",
      "3                    43                  3                 15   \n",
      "4                     0                  0                 60   \n",
      "...                 ...                ...                ...   \n",
      "6194                433               5426               9006   \n",
      "6195                571               3307               4737   \n",
      "6196               8483              24028              30266   \n",
      "6197               6430              16099              20215   \n",
      "6198                  0                  6                  4   \n",
      "\n",
      "      reads_SRR23891366  reads_SRR23891367  reads_SRR23891370  \\\n",
      "0                     0                 42                  3   \n",
      "1                     0                 31                  0   \n",
      "2                     0                  2                  0   \n",
      "3                     0                  6                 18   \n",
      "4                   260                  0                  0   \n",
      "...                 ...                ...                ...   \n",
      "6194              11286                417                377   \n",
      "6195               7091                472                284   \n",
      "6196              45841               5386               2494   \n",
      "6197              26409               2784               1612   \n",
      "6198                  0                  0                  0   \n",
      "\n",
      "      reads_SRR23891372  \n",
      "0                    13  \n",
      "1                    29  \n",
      "2                     0  \n",
      "3                    54  \n",
      "4                     4  \n",
      "...                 ...  \n",
      "6194              17117  \n",
      "6195               6358  \n",
      "6196              63056  \n",
      "6197              43094  \n",
      "6198                 14  \n",
      "\n",
      "[6199 rows x 17 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Read the first CSV file\n",
    "file1_path = 'no_length/reads_gene/counts_sum_nohapAB.csv'\n",
    "df1 = pd.read_csv(file1_path, sep='\\t')\n",
    "\n",
    "# Read the second CSV file\n",
    "file2_path = 'no_length/reads_gene/counts_sum.csv'\n",
    "df2 = pd.read_csv(file2_path, sep='\\t')\n",
    "\n",
    "# Concatenate the dataframes vertically\n",
    "result_df = pd.concat([df2, df1], ignore_index=True, sort=False)\n",
    "\n",
    "result_df.to_csv('no_length/reads_gene/counts_sum1.csv', sep='\\t', index=False)\n",
    "\n",
    "# Display the result or save it to a new CSV file\n",
    "print(result_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now there can also be some geneids existing just for hapA or hapB. We want these gene_ids as well. :-\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "dfs_hapA = []\n",
    "dfs_hapB = []                        ## By doing this we are making a two different csv files which will have hapA and hapB geneids respectively.\n",
    "\n",
    "for accession_number in accession_numbers:\n",
    "    df1=pd.read_csv(f'no_length/reads_gene/hapA/csv/counts_hapA_{accession_number}.csv',sep=\"\\t\")\n",
    "    df2=pd.read_csv(f'no_length/reads_gene/hapB/csv/counts_hapB_{accession_number}.csv',sep=\"\\t\")\n",
    "\n",
    "    df1[\"Geneid\"] = df1[\"Geneid\"].str.slice(0, -2)\n",
    "    df2[\"Geneid\"] = df2[\"Geneid\"].str.slice(0, -2)\n",
    "\n",
    "    df_lenA=df1[[\"Geneid\", \"Length\"]]\n",
    "    df_lenB=df2[[\"Geneid\", \"Length\"]]\n",
    "    # Extract the columns Geneid and reads\n",
    "    df1 = df1[[\"Geneid\", f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\"]]\n",
    "    df2 = df2[[\"Geneid\", f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\"]]\n",
    "\n",
    "    # Rename the reads column\n",
    "    df1 = df1.rename(columns={f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\": f\"reads_{accession_number}\"})\n",
    "    df2 = df2.rename(columns={f\"no_length/mapped_mRNA/bams_collapsed/{accession_number}.map_mRNAAligned.sortedByCoord.out.bam\": f\"reads_{accession_number}\"})\n",
    "\n",
    "    dfs_hapA.append(df1)\n",
    "    dfs_hapB.append(df2)\n",
    "\n",
    "# Concatenate DataFrames for hapA and hapB\n",
    "df_hapA = pd.concat(dfs_hapA, axis=1)\n",
    "df_hapB = pd.concat(dfs_hapB, axis=1)\n",
    "\n",
    "df_hapA = df_hapA.loc[:, ~df_hapA.columns.duplicated()]           ## removing duplicated columns\n",
    "df_hapB = df_hapB.loc[:, ~df_hapB.columns.duplicated()]\n",
    "\n",
    "df_hapA.to_csv(\"no_length/reads_gene/counts_sumA.csv\", sep=\"\\t\", index=False)\n",
    "df_hapB.to_csv(\"no_length/reads_gene/counts_sumB.csv\", sep=\"\\t\", index=False)\n",
    "\n",
    "df_lenA.to_csv(\"no_length/reads_gene/length_A.csv\", sep=\"\\t\", index=False)\n",
    "df_lenB.to_csv(\"no_length/reads_gene/length_B.csv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>reads_SRR23891364</th>\n",
       "      <th>reads_SRR23891365</th>\n",
       "      <th>reads_SRR23891368</th>\n",
       "      <th>reads_SRR23891369</th>\n",
       "      <th>reads_SRR23891371</th>\n",
       "      <th>reads_SRR23891373</th>\n",
       "      <th>reads_SRR23891374</th>\n",
       "      <th>reads_SRR23891375</th>\n",
       "      <th>reads_SRR23891376</th>\n",
       "      <th>reads_SRR23891377</th>\n",
       "      <th>reads_SRR23891378</th>\n",
       "      <th>reads_SRR23891363</th>\n",
       "      <th>reads_SRR23891366</th>\n",
       "      <th>reads_SRR23891367</th>\n",
       "      <th>reads_SRR23891370</th>\n",
       "      <th>reads_SRR23891372</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C1_00010W</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>2</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>48</td>\n",
       "      <td>41</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C1_00020C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C1_00030C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C1_00040W</td>\n",
       "      <td>67</td>\n",
       "      <td>113</td>\n",
       "      <td>40</td>\n",
       "      <td>41</td>\n",
       "      <td>67</td>\n",
       "      <td>75</td>\n",
       "      <td>25</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>42</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C1_00050C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>260</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6180</th>\n",
       "      <td>CR_10820W</td>\n",
       "      <td>631</td>\n",
       "      <td>236</td>\n",
       "      <td>258</td>\n",
       "      <td>26</td>\n",
       "      <td>249</td>\n",
       "      <td>306</td>\n",
       "      <td>129</td>\n",
       "      <td>103</td>\n",
       "      <td>50</td>\n",
       "      <td>41</td>\n",
       "      <td>57</td>\n",
       "      <td>356</td>\n",
       "      <td>172</td>\n",
       "      <td>131</td>\n",
       "      <td>46</td>\n",
       "      <td>306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6181</th>\n",
       "      <td>CR_10830C</td>\n",
       "      <td>753</td>\n",
       "      <td>445</td>\n",
       "      <td>872</td>\n",
       "      <td>1484</td>\n",
       "      <td>651</td>\n",
       "      <td>258</td>\n",
       "      <td>479</td>\n",
       "      <td>1756</td>\n",
       "      <td>985</td>\n",
       "      <td>907</td>\n",
       "      <td>292</td>\n",
       "      <td>548</td>\n",
       "      <td>0</td>\n",
       "      <td>537</td>\n",
       "      <td>306</td>\n",
       "      <td>592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6182</th>\n",
       "      <td>CR_10840C</td>\n",
       "      <td>19494</td>\n",
       "      <td>10371</td>\n",
       "      <td>24680</td>\n",
       "      <td>6560</td>\n",
       "      <td>21799</td>\n",
       "      <td>16328</td>\n",
       "      <td>24314</td>\n",
       "      <td>19324</td>\n",
       "      <td>25536</td>\n",
       "      <td>13230</td>\n",
       "      <td>14085</td>\n",
       "      <td>10882</td>\n",
       "      <td>25041</td>\n",
       "      <td>5675</td>\n",
       "      <td>9923</td>\n",
       "      <td>19744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183</th>\n",
       "      <td>CR_10850C</td>\n",
       "      <td>7766</td>\n",
       "      <td>6711</td>\n",
       "      <td>10454</td>\n",
       "      <td>9884</td>\n",
       "      <td>10842</td>\n",
       "      <td>11125</td>\n",
       "      <td>11011</td>\n",
       "      <td>8051</td>\n",
       "      <td>9387</td>\n",
       "      <td>7149</td>\n",
       "      <td>5004</td>\n",
       "      <td>7287</td>\n",
       "      <td>6349</td>\n",
       "      <td>4522</td>\n",
       "      <td>2708</td>\n",
       "      <td>11150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6184</th>\n",
       "      <td>CR_10860C</td>\n",
       "      <td>250</td>\n",
       "      <td>338</td>\n",
       "      <td>1315</td>\n",
       "      <td>1084</td>\n",
       "      <td>1021</td>\n",
       "      <td>494</td>\n",
       "      <td>599</td>\n",
       "      <td>1115</td>\n",
       "      <td>1495</td>\n",
       "      <td>945</td>\n",
       "      <td>522</td>\n",
       "      <td>224</td>\n",
       "      <td>2</td>\n",
       "      <td>188</td>\n",
       "      <td>225</td>\n",
       "      <td>593</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6185 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Geneid  reads_SRR23891364  reads_SRR23891365  reads_SRR23891368  \\\n",
       "0     C1_00010W                  0                  0                 72   \n",
       "1     C1_00020C                  0                  0                 12   \n",
       "2     C1_00030C                  0                  0                  0   \n",
       "3     C1_00040W                 67                113                 40   \n",
       "4     C1_00050C                  0                  0                  5   \n",
       "...         ...                ...                ...                ...   \n",
       "6180  CR_10820W                631                236                258   \n",
       "6181  CR_10830C                753                445                872   \n",
       "6182  CR_10840C              19494              10371              24680   \n",
       "6183  CR_10850C               7766               6711              10454   \n",
       "6184  CR_10860C                250                338               1315   \n",
       "\n",
       "      reads_SRR23891369  reads_SRR23891371  reads_SRR23891373  \\\n",
       "0                     2                 46                  0   \n",
       "1                     0                  0                  0   \n",
       "2                     0                  0                  0   \n",
       "3                    41                 67                 75   \n",
       "4                     0                  1                  3   \n",
       "...                 ...                ...                ...   \n",
       "6180                 26                249                306   \n",
       "6181               1484                651                258   \n",
       "6182               6560              21799              16328   \n",
       "6183               9884              10842              11125   \n",
       "6184               1084               1021                494   \n",
       "\n",
       "      reads_SRR23891374  reads_SRR23891375  reads_SRR23891376  \\\n",
       "0                    14                 48                 41   \n",
       "1                    22                  0                  1   \n",
       "2                     0                  0                  1   \n",
       "3                    25                  5                  5   \n",
       "4                     0                  4                  3   \n",
       "...                 ...                ...                ...   \n",
       "6180                129                103                 50   \n",
       "6181                479               1756                985   \n",
       "6182              24314              19324              25536   \n",
       "6183              11011               8051               9387   \n",
       "6184                599               1115               1495   \n",
       "\n",
       "      reads_SRR23891377  reads_SRR23891378  reads_SRR23891363  \\\n",
       "0                     7                  8                  0   \n",
       "1                     2                  1                  0   \n",
       "2                     3                  1                  0   \n",
       "3                    42                  3                 15   \n",
       "4                     0                  0                 60   \n",
       "...                 ...                ...                ...   \n",
       "6180                 41                 57                356   \n",
       "6181                907                292                548   \n",
       "6182              13230              14085              10882   \n",
       "6183               7149               5004               7287   \n",
       "6184                945                522                224   \n",
       "\n",
       "      reads_SRR23891366  reads_SRR23891367  reads_SRR23891370  \\\n",
       "0                     0                 42                  3   \n",
       "1                     0                 31                  0   \n",
       "2                     0                  1                  0   \n",
       "3                     0                  1                 18   \n",
       "4                   260                  0                  0   \n",
       "...                 ...                ...                ...   \n",
       "6180                172                131                 46   \n",
       "6181                  0                537                306   \n",
       "6182              25041               5675               9923   \n",
       "6183               6349               4522               2708   \n",
       "6184                  2                188                225   \n",
       "\n",
       "      reads_SRR23891372  \n",
       "0                    10  \n",
       "1                    29  \n",
       "2                     0  \n",
       "3                    54  \n",
       "4                     4  \n",
       "...                 ...  \n",
       "6180                306  \n",
       "6181                592  \n",
       "6182              19744  \n",
       "6183              11150  \n",
       "6184                593  \n",
       "\n",
       "[6185 rows x 17 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hapA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "from the above csv files we will extract the geneids of hapB which are not present in hapA and vice versa."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hapA=pd.read_csv(\"no_length/reads_gene/counts_sumA.csv\", sep='\\t')\n",
    "df_hapB=pd.read_csv(\"no_length/reads_gene/counts_sumB.csv\", sep='\\t')\n",
    "\n",
    "diff1= df_hapA[~df_hapA[\"Geneid\"].isin(df_hapB[\"Geneid\"])]\n",
    "diff2= df_hapB[~df_hapB[\"Geneid\"].isin(df_hapA[\"Geneid\"])]\n",
    "\n",
    "diff1 = diff1.loc[:, ~diff1.columns.duplicated()]\n",
    "diff2 = diff2.loc[:, ~diff2.columns.duplicated()]\n",
    "\n",
    "diff1.to_csv(\"no_length/reads_gene/diff1.csv\", sep=\"\\t\", index=False)\n",
    "diff2.to_csv(\"no_length/reads_gene/diff2.csv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>reads_SRR23891364</th>\n",
       "      <th>reads_SRR23891365</th>\n",
       "      <th>reads_SRR23891368</th>\n",
       "      <th>reads_SRR23891369</th>\n",
       "      <th>reads_SRR23891371</th>\n",
       "      <th>reads_SRR23891373</th>\n",
       "      <th>reads_SRR23891374</th>\n",
       "      <th>reads_SRR23891375</th>\n",
       "      <th>reads_SRR23891376</th>\n",
       "      <th>reads_SRR23891377</th>\n",
       "      <th>reads_SRR23891378</th>\n",
       "      <th>reads_SRR23891363</th>\n",
       "      <th>reads_SRR23891366</th>\n",
       "      <th>reads_SRR23891367</th>\n",
       "      <th>reads_SRR23891370</th>\n",
       "      <th>reads_SRR23891372</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3986</th>\n",
       "      <td>C5_01740C</td>\n",
       "      <td>379</td>\n",
       "      <td>165</td>\n",
       "      <td>227</td>\n",
       "      <td>649</td>\n",
       "      <td>411</td>\n",
       "      <td>501</td>\n",
       "      <td>527</td>\n",
       "      <td>591</td>\n",
       "      <td>331</td>\n",
       "      <td>391</td>\n",
       "      <td>194</td>\n",
       "      <td>207</td>\n",
       "      <td>8</td>\n",
       "      <td>142</td>\n",
       "      <td>35</td>\n",
       "      <td>555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3987</th>\n",
       "      <td>C5_01750C</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>26</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3988</th>\n",
       "      <td>C5_01760C</td>\n",
       "      <td>1086</td>\n",
       "      <td>498</td>\n",
       "      <td>524</td>\n",
       "      <td>761</td>\n",
       "      <td>1480</td>\n",
       "      <td>1588</td>\n",
       "      <td>1361</td>\n",
       "      <td>1558</td>\n",
       "      <td>1459</td>\n",
       "      <td>691</td>\n",
       "      <td>580</td>\n",
       "      <td>787</td>\n",
       "      <td>2784</td>\n",
       "      <td>520</td>\n",
       "      <td>181</td>\n",
       "      <td>1425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3989</th>\n",
       "      <td>C5_01770C</td>\n",
       "      <td>3746</td>\n",
       "      <td>3037</td>\n",
       "      <td>1934</td>\n",
       "      <td>1782</td>\n",
       "      <td>3566</td>\n",
       "      <td>4422</td>\n",
       "      <td>3889</td>\n",
       "      <td>1487</td>\n",
       "      <td>2260</td>\n",
       "      <td>1691</td>\n",
       "      <td>1722</td>\n",
       "      <td>2580</td>\n",
       "      <td>3892</td>\n",
       "      <td>542</td>\n",
       "      <td>688</td>\n",
       "      <td>4158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3990</th>\n",
       "      <td>C5_01780W</td>\n",
       "      <td>1354</td>\n",
       "      <td>830</td>\n",
       "      <td>467</td>\n",
       "      <td>80</td>\n",
       "      <td>1097</td>\n",
       "      <td>1152</td>\n",
       "      <td>1087</td>\n",
       "      <td>758</td>\n",
       "      <td>232</td>\n",
       "      <td>163</td>\n",
       "      <td>320</td>\n",
       "      <td>1043</td>\n",
       "      <td>2926</td>\n",
       "      <td>267</td>\n",
       "      <td>148</td>\n",
       "      <td>1304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3991</th>\n",
       "      <td>C5_01790C</td>\n",
       "      <td>139</td>\n",
       "      <td>15</td>\n",
       "      <td>134</td>\n",
       "      <td>142</td>\n",
       "      <td>35</td>\n",
       "      <td>43</td>\n",
       "      <td>129</td>\n",
       "      <td>179</td>\n",
       "      <td>58</td>\n",
       "      <td>26</td>\n",
       "      <td>46</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>25</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Geneid  reads_SRR23891364  reads_SRR23891365  reads_SRR23891368  \\\n",
       "3986  C5_01740C                379                165                227   \n",
       "3987  C5_01750C                  0                 22                  0   \n",
       "3988  C5_01760C               1086                498                524   \n",
       "3989  C5_01770C               3746               3037               1934   \n",
       "3990  C5_01780W               1354                830                467   \n",
       "3991  C5_01790C                139                 15                134   \n",
       "\n",
       "      reads_SRR23891369  reads_SRR23891371  reads_SRR23891373  \\\n",
       "3986                649                411                501   \n",
       "3987                  1                  0                 25   \n",
       "3988                761               1480               1588   \n",
       "3989               1782               3566               4422   \n",
       "3990                 80               1097               1152   \n",
       "3991                142                 35                 43   \n",
       "\n",
       "      reads_SRR23891374  reads_SRR23891375  reads_SRR23891376  \\\n",
       "3986                527                591                331   \n",
       "3987                 26                  3                  0   \n",
       "3988               1361               1558               1459   \n",
       "3989               3889               1487               2260   \n",
       "3990               1087                758                232   \n",
       "3991                129                179                 58   \n",
       "\n",
       "      reads_SRR23891377  reads_SRR23891378  reads_SRR23891363  \\\n",
       "3986                391                194                207   \n",
       "3987                  0                  9                  6   \n",
       "3988                691                580                787   \n",
       "3989               1691               1722               2580   \n",
       "3990                163                320               1043   \n",
       "3991                 26                 46                 19   \n",
       "\n",
       "      reads_SRR23891366  reads_SRR23891367  reads_SRR23891370  \\\n",
       "3986                  8                142                 35   \n",
       "3987                  1                  6                  1   \n",
       "3988               2784                520                181   \n",
       "3989               3892                542                688   \n",
       "3990               2926                267                148   \n",
       "3991                  1                190                 25   \n",
       "\n",
       "      reads_SRR23891372  \n",
       "3986                555  \n",
       "3987                  3  \n",
       "3988               1425  \n",
       "3989               4158  \n",
       "3990               1304  \n",
       "3991                105  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff1      ## geneids of hapA which are not there in hapB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>reads_SRR23891364</th>\n",
       "      <th>reads_SRR23891365</th>\n",
       "      <th>reads_SRR23891368</th>\n",
       "      <th>reads_SRR23891369</th>\n",
       "      <th>reads_SRR23891371</th>\n",
       "      <th>reads_SRR23891373</th>\n",
       "      <th>reads_SRR23891374</th>\n",
       "      <th>reads_SRR23891375</th>\n",
       "      <th>reads_SRR23891376</th>\n",
       "      <th>reads_SRR23891377</th>\n",
       "      <th>reads_SRR23891378</th>\n",
       "      <th>reads_SRR23891363</th>\n",
       "      <th>reads_SRR23891366</th>\n",
       "      <th>reads_SRR23891367</th>\n",
       "      <th>reads_SRR23891370</th>\n",
       "      <th>reads_SRR23891372</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3986</th>\n",
       "      <td>C5_01745W</td>\n",
       "      <td>863</td>\n",
       "      <td>827</td>\n",
       "      <td>465</td>\n",
       "      <td>529</td>\n",
       "      <td>668</td>\n",
       "      <td>1021</td>\n",
       "      <td>580</td>\n",
       "      <td>720</td>\n",
       "      <td>861</td>\n",
       "      <td>431</td>\n",
       "      <td>283</td>\n",
       "      <td>533</td>\n",
       "      <td>57</td>\n",
       "      <td>184</td>\n",
       "      <td>239</td>\n",
       "      <td>778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3987</th>\n",
       "      <td>C5_01755C</td>\n",
       "      <td>35</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "      <td>58</td>\n",
       "      <td>36</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3988</th>\n",
       "      <td>C5_01765C</td>\n",
       "      <td>1521</td>\n",
       "      <td>1228</td>\n",
       "      <td>856</td>\n",
       "      <td>1089</td>\n",
       "      <td>1098</td>\n",
       "      <td>1999</td>\n",
       "      <td>1365</td>\n",
       "      <td>1055</td>\n",
       "      <td>1255</td>\n",
       "      <td>1035</td>\n",
       "      <td>738</td>\n",
       "      <td>1204</td>\n",
       "      <td>1719</td>\n",
       "      <td>424</td>\n",
       "      <td>243</td>\n",
       "      <td>1608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3989</th>\n",
       "      <td>C5_01775C</td>\n",
       "      <td>947</td>\n",
       "      <td>474</td>\n",
       "      <td>621</td>\n",
       "      <td>182</td>\n",
       "      <td>914</td>\n",
       "      <td>1058</td>\n",
       "      <td>703</td>\n",
       "      <td>833</td>\n",
       "      <td>886</td>\n",
       "      <td>835</td>\n",
       "      <td>364</td>\n",
       "      <td>386</td>\n",
       "      <td>597</td>\n",
       "      <td>193</td>\n",
       "      <td>289</td>\n",
       "      <td>1154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3990</th>\n",
       "      <td>C5_01785W</td>\n",
       "      <td>343</td>\n",
       "      <td>211</td>\n",
       "      <td>182</td>\n",
       "      <td>14</td>\n",
       "      <td>242</td>\n",
       "      <td>518</td>\n",
       "      <td>353</td>\n",
       "      <td>345</td>\n",
       "      <td>222</td>\n",
       "      <td>113</td>\n",
       "      <td>155</td>\n",
       "      <td>217</td>\n",
       "      <td>219</td>\n",
       "      <td>30</td>\n",
       "      <td>69</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Geneid  reads_SRR23891364  reads_SRR23891365  reads_SRR23891368  \\\n",
       "3986  C5_01745W                863                827                465   \n",
       "3987  C5_01755C                 35                 22                  2   \n",
       "3988  C5_01765C               1521               1228                856   \n",
       "3989  C5_01775C                947                474                621   \n",
       "3990  C5_01785W                343                211                182   \n",
       "\n",
       "      reads_SRR23891369  reads_SRR23891371  reads_SRR23891373  \\\n",
       "3986                529                668               1021   \n",
       "3987                  2                 29                 58   \n",
       "3988               1089               1098               1999   \n",
       "3989                182                914               1058   \n",
       "3990                 14                242                518   \n",
       "\n",
       "      reads_SRR23891374  reads_SRR23891375  reads_SRR23891376  \\\n",
       "3986                580                720                861   \n",
       "3987                 36                 12                  7   \n",
       "3988               1365               1055               1255   \n",
       "3989                703                833                886   \n",
       "3990                353                345                222   \n",
       "\n",
       "      reads_SRR23891377  reads_SRR23891378  reads_SRR23891363  \\\n",
       "3986                431                283                533   \n",
       "3987                  1                 17                 95   \n",
       "3988               1035                738               1204   \n",
       "3989                835                364                386   \n",
       "3990                113                155                217   \n",
       "\n",
       "      reads_SRR23891366  reads_SRR23891367  reads_SRR23891370  \\\n",
       "3986                 57                184                239   \n",
       "3987                  0                  1                  4   \n",
       "3988               1719                424                243   \n",
       "3989                597                193                289   \n",
       "3990                219                 30                 69   \n",
       "\n",
       "      reads_SRR23891372  \n",
       "3986                778  \n",
       "3987                 36  \n",
       "3988               1608  \n",
       "3989               1154  \n",
       "3990                334  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff2       ## geneids of hapB which are not there in hapA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will concatinate the diff1 and diff2 with the counts_sum1.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "final= pd.read_csv(\"no_length/reads_gene/counts_sum1.csv\", sep=\"\\t\")\n",
    "diff2= pd.read_csv(\"no_length/reads_gene/diff2.csv\", sep=\"\\t\")\n",
    "diff1= pd.read_csv(\"no_length/reads_gene/diff1.csv\", sep=\"\\t\")\n",
    "dfs=[]\n",
    "dfs.append(final)\n",
    "dfs.append(diff1)\n",
    "dfs.append(diff2)\n",
    "\n",
    "final_counts_sum= pd.concat(dfs, axis=0)   # axis 0 means concatinating along y axis or vertically"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>reads_SRR23891364</th>\n",
       "      <th>reads_SRR23891365</th>\n",
       "      <th>reads_SRR23891368</th>\n",
       "      <th>reads_SRR23891369</th>\n",
       "      <th>reads_SRR23891371</th>\n",
       "      <th>reads_SRR23891373</th>\n",
       "      <th>reads_SRR23891374</th>\n",
       "      <th>reads_SRR23891375</th>\n",
       "      <th>reads_SRR23891376</th>\n",
       "      <th>reads_SRR23891377</th>\n",
       "      <th>reads_SRR23891378</th>\n",
       "      <th>reads_SRR23891363</th>\n",
       "      <th>reads_SRR23891366</th>\n",
       "      <th>reads_SRR23891367</th>\n",
       "      <th>reads_SRR23891370</th>\n",
       "      <th>reads_SRR23891372</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C1_00010W</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>2</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>48</td>\n",
       "      <td>42</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C1_00020C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C1_00030C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C1_00040W</td>\n",
       "      <td>67</td>\n",
       "      <td>113</td>\n",
       "      <td>40</td>\n",
       "      <td>41</td>\n",
       "      <td>67</td>\n",
       "      <td>75</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C1_00050C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>260</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C5_01745W</td>\n",
       "      <td>863</td>\n",
       "      <td>827</td>\n",
       "      <td>465</td>\n",
       "      <td>529</td>\n",
       "      <td>668</td>\n",
       "      <td>1021</td>\n",
       "      <td>580</td>\n",
       "      <td>720</td>\n",
       "      <td>861</td>\n",
       "      <td>431</td>\n",
       "      <td>283</td>\n",
       "      <td>533</td>\n",
       "      <td>57</td>\n",
       "      <td>184</td>\n",
       "      <td>239</td>\n",
       "      <td>778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C5_01755C</td>\n",
       "      <td>35</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "      <td>58</td>\n",
       "      <td>36</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C5_01765C</td>\n",
       "      <td>1521</td>\n",
       "      <td>1228</td>\n",
       "      <td>856</td>\n",
       "      <td>1089</td>\n",
       "      <td>1098</td>\n",
       "      <td>1999</td>\n",
       "      <td>1365</td>\n",
       "      <td>1055</td>\n",
       "      <td>1255</td>\n",
       "      <td>1035</td>\n",
       "      <td>738</td>\n",
       "      <td>1204</td>\n",
       "      <td>1719</td>\n",
       "      <td>424</td>\n",
       "      <td>243</td>\n",
       "      <td>1608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C5_01775C</td>\n",
       "      <td>947</td>\n",
       "      <td>474</td>\n",
       "      <td>621</td>\n",
       "      <td>182</td>\n",
       "      <td>914</td>\n",
       "      <td>1058</td>\n",
       "      <td>703</td>\n",
       "      <td>833</td>\n",
       "      <td>886</td>\n",
       "      <td>835</td>\n",
       "      <td>364</td>\n",
       "      <td>386</td>\n",
       "      <td>597</td>\n",
       "      <td>193</td>\n",
       "      <td>289</td>\n",
       "      <td>1154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C5_01785W</td>\n",
       "      <td>343</td>\n",
       "      <td>211</td>\n",
       "      <td>182</td>\n",
       "      <td>14</td>\n",
       "      <td>242</td>\n",
       "      <td>518</td>\n",
       "      <td>353</td>\n",
       "      <td>345</td>\n",
       "      <td>222</td>\n",
       "      <td>113</td>\n",
       "      <td>155</td>\n",
       "      <td>217</td>\n",
       "      <td>219</td>\n",
       "      <td>30</td>\n",
       "      <td>69</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6210 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Geneid  reads_SRR23891364  reads_SRR23891365  reads_SRR23891368  \\\n",
       "0   C1_00010W                  0                  0                 81   \n",
       "1   C1_00020C                  0                  0                 12   \n",
       "2   C1_00030C                  0                  0                  0   \n",
       "3   C1_00040W                 67                113                 40   \n",
       "4   C1_00050C                  0                  0                  5   \n",
       "..        ...                ...                ...                ...   \n",
       "0   C5_01745W                863                827                465   \n",
       "1   C5_01755C                 35                 22                  2   \n",
       "2   C5_01765C               1521               1228                856   \n",
       "3   C5_01775C                947                474                621   \n",
       "4   C5_01785W                343                211                182   \n",
       "\n",
       "    reads_SRR23891369  reads_SRR23891371  reads_SRR23891373  \\\n",
       "0                   2                 46                  0   \n",
       "1                   0                  0                  1   \n",
       "2                   0                  0                  0   \n",
       "3                  41                 67                 75   \n",
       "4                   0                  1                  3   \n",
       "..                ...                ...                ...   \n",
       "0                 529                668               1021   \n",
       "1                   2                 29                 58   \n",
       "2                1089               1098               1999   \n",
       "3                 182                914               1058   \n",
       "4                  14                242                518   \n",
       "\n",
       "    reads_SRR23891374  reads_SRR23891375  reads_SRR23891376  \\\n",
       "0                  14                 48                 42   \n",
       "1                  22                  0                  1   \n",
       "2                   0                  0                  1   \n",
       "3                  26                  5                  5   \n",
       "4                   0                  4                  3   \n",
       "..                ...                ...                ...   \n",
       "0                 580                720                861   \n",
       "1                  36                 12                  7   \n",
       "2                1365               1055               1255   \n",
       "3                 703                833                886   \n",
       "4                 353                345                222   \n",
       "\n",
       "    reads_SRR23891377  reads_SRR23891378  reads_SRR23891363  \\\n",
       "0                   7                  8                  8   \n",
       "1                   2                  1                  0   \n",
       "2                   3                  1                  0   \n",
       "3                  43                  3                 15   \n",
       "4                   0                  0                 60   \n",
       "..                ...                ...                ...   \n",
       "0                 431                283                533   \n",
       "1                   1                 17                 95   \n",
       "2                1035                738               1204   \n",
       "3                 835                364                386   \n",
       "4                 113                155                217   \n",
       "\n",
       "    reads_SRR23891366  reads_SRR23891367  reads_SRR23891370  reads_SRR23891372  \n",
       "0                   0                 42                  3                 13  \n",
       "1                   0                 31                  0                 29  \n",
       "2                   0                  2                  0                  0  \n",
       "3                   0                  6                 18                 54  \n",
       "4                 260                  0                  0                  4  \n",
       "..                ...                ...                ...                ...  \n",
       "0                  57                184                239                778  \n",
       "1                   0                  1                  4                 36  \n",
       "2                1719                424                243               1608  \n",
       "3                 597                193                289               1154  \n",
       "4                 219                 30                 69                334  \n",
       "\n",
       "[6210 rows x 17 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_counts_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_counts_sum.to_csv(\"no_length/reads_gene/final_counts_sum.csv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remvoing unneccesary csv files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm no_length/reads_gene/counts_sum1.csv\n",
    "!rm no_length/reads_gene/counts_sum.csv\n",
    "!rm no_length/reads_gene/diff1.csv\n",
    "!rm no_length/reads_gene/diff2.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we finaaly have our reads per gne counts with us. With this we can now perform DeSeq and TE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "## This code is for creating the final Length file which will have length of all the gneids which we can further use for the deseq and te anysis\n",
    "len_dfs=[]\n",
    "\n",
    "lenA= pd.read_csv(\"no_length/reads_gene/length_A.csv\", sep='\\t')\n",
    "lenB= pd.read_csv(\"no_length/reads_gene/length_B.csv\", sep='\\t')\n",
    "len_noAB=pd.read_csv(\"no_length/reads_gene/length_noAB.csv\", sep='\\t')\n",
    "\n",
    "len_dfs.append(lenA)\n",
    "len_dfs.append(lenB)\n",
    "len_dfs.append(len_noAB)\n",
    "\n",
    "final_len= pd.concat(len_dfs, axis=0)\n",
    "final_len=final_len.drop_duplicates(subset=[\"Geneid\"])\n",
    "final_len.to_csv(\"no_length/reads_gene/final_len.csv\", sep=\"\\t\", index=False)\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
